scripts/update-changelog.js

```diff
@@ -0,0 +1,77 @@
+/**
+ * This script updates changelogs after lerna has updated versions in the respective areas (packages/*, experimental/packages/*)
+ * - removes all empty subsections (bugs, enhancements, etc.) in the changelog.
+ * - replaces the "Unreleased"-header with the version from the first non-private package in the directory (versions are expected to be uniform across a changelog)
+ * - adds a new "Unreleased"-header with empty subsections at the top
+ *
+ * Usage (from project root):
+ * - node scripts/update-changelog.js [PATH TO CHANGELOG] [DIRECTORY CONTAINING ASSOCIATED PACKAGES]
+ * Examples:
+ * - node scripts/update-changelog.js ./CHANGELOG.md ./packages
+ * - node scripts/update-changelog.js ./experimental/CHANGELOG.md ./experimental/packages
+ */
+
+const fs = require('fs');
+const path = require("path");
+
+const EMPTY_UNRELEASED_SECTION = `## Unreleased
+
+### :boom: Breaking Change
+
+### :rocket: (Enhancement)
+
+### :bug: (Bug Fix)
+
+### :books: (Refine Doc)
+
+### :house: (Internal)
+
+`
+
+function findFirstPackageVersion(basePath){
+  const packageDirs = fs.readdirSync(basePath);
+  for(const packageDir of packageDirs){
+    const packageJsonPath = path.join(basePath, packageDir, 'package.json');
+    try {
+      const packageJson =  JSON.parse(fs.readFileSync(packageJsonPath, 'utf-8'));
+
+      if(packageJson.private === true || packageJson.private === 'true'){
+        console.log('Skipping version from private package at', packageJsonPath);
+        continue;
+      }
+
+      if(packageJson.version != null){
+        return version;
+      }
+
+      console.log('Version in', packageJsonPath, 'was null or undefined, skipping');
+    } catch (err) {
+      console.log('Could not get package JSON', packageJsonPath, err);
+    }
+  }
+  throw new Error('Unable to extract version from packages in ' + basePath);
+}
+
+function replaceEmptySection(changelog){
+  // Only match ## at the end in case the last header does not have any entries and the next one is '## version',
+  // this makes it safe to replace with only '##'
+  return changelog.replace(RegExp('###.*\n*##', 'gm'), '##');
```

Comment:
Some nits:

- Could anchor those section headers at the start of lines to avoid an (unlikely) surprise: `'^###.*\n*^##'`
- You could put the trailing `##` in a [lookahead assertion](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_expressions/Assertions#other_assertions) which says "assert that this thing matches next, but don't include it in the match string". The result is that all the matches of empty sections will be non-overlapping, so the 'g' flag will replace them all in the single `changelog.replace(...)` call and you'll no longer need the `while (...)`-loop. If you do this, the replacement string should now be the empty string.

```js
const trimmed = changelog.replace(new RegExp('^###.*\n*(?=^##)', 'gm'), '');
```

Reply:
Ah yes this is way better, great suggestion!
Applied in 90b3977b57cedcd978504da68d8430b196729e6b. Thanks! :slightly_smiling_face: